qiskit_nature/algorithms/pes_samplers/bopes_sampler.py

```diff
@@ -113,6 +113,13 @@ def sample(self, problem: BaseProblem, points: List[float]) -> BOPESSamplerResul
         """
         self._problem = problem
         self._driver = problem.driver
+        # We have to force the creation of the solver so that we work on the same solver
+        # instance before and after _gss.solve
+        self._gss.prepare_solve(problem, None)
+        if isinstance(self._gss.solver, VariationalAlgorithm):  # type: ignore
```

Comment:
There are various of these `isinstance` checks in the `BopesSampler` code. This is what the culprit of the problem appears to be, because this check will fail when `VQEUCCFactory` is being used. Thus, I suggest you try to extend this check instead, similar to the following:
```suggestion
        if isinstance(self._gss.solver, (VariationalAlgorithm, VQEUCCFactory)):  # type: ignore
```
If you do this everywhere, that should also fix the problem.

I am deliberately suggesting to use `VQEUCCFactory` here, because `MinimumEigensolverFactory` would also contain the `NumPyMinimumEigensolverFactory` (which we do not want).
@pbark does the `BopesSampler` also support `VibrationalStructureProblems`? If so, we will also need to support the `VQEUVCCFactory`.

Reply:
@mrossinek  In BopesSampler we are manipulating internal structure of members of GroundStateEigensolver. This breaks the encapsulation and produces this hard to extend design. Also, the method solve in GroundStateEigensolver processes the problem, creates solver instance and performs the calculation. This breaks single responsibility principle. When SOLID principles are violated, extension is hard and produces even worst design. I don't think that the solution is to put in this class if code everywhere for internal structure of other classes. We have to allow GroundStateEigensolver implementations to vary and BOPESSampler must still work.

Reply:
I see what you're saying. However, in that case you should also change the `_run_single_point` method to remove the overhead of `prepare_solve` being called twice and instead call `compute_minimum_eigenvalue` and `problem.interpret` directly.

Reply:
@mrossinek You're right. I'll check.

Reply:
I wonder if we should already add the `is_variational` utility to the `GroundStateSolver` which I proposed in #384 :thinking: 
What do you think?